File: ApexFit/Models/SleepSession.swift
import Foundation
import SwiftData

@Model
final class SleepSession {
    var id: UUID
    var startDate: Date
    var endDate: Date
    var isMainSleep: Bool
    var isNap: Bool
    var totalSleepMinutes: Double
    var timeInBedMinutes: Double
    var lightSleepMinutes: Double
    var deepSleepMinutes: Double
    var remSleepMinutes: Double
    var awakeMinutes: Double
    var awakenings: Int
    var sleepOnsetLatencyMinutes: Double?
    var sleepEfficiency: Double
    var sleepPerformance: Double?
    var sleepNeedHours: Double?
    var createdAt: Date

    var dailyMetric: DailyMetric?

    @Relationship(deleteRule: .cascade, inverse: \SleepStage.sleepSession)
    var stages: [SleepStage]

    init(startDate: Date, endDate: Date, isMainSleep: Bool = true) {
        self.id = UUID()
        self.startDate = startDate
        self.endDate = endDate
        self.isMainSleep = isMainSleep
        self.isNap = !isMainSleep
        self.totalSleepMinutes = 0
        self.timeInBedMinutes = endDate.timeIntervalSince(startDate) / 60.0
        self.lightSleepMinutes = 0
        self.deepSleepMinutes = 0
        self.remSleepMinutes = 0
        self.awakeMinutes = 0
        self.awakenings = 0
        self.sleepEfficiency = 0
        self.createdAt = Date()
        self.stages = []
    }

    var totalSleepHours: Double {
        totalSleepMinutes / 60.0
    }

    var deepSleepPercentage: Double {
        guard totalSleepMinutes > 0 else { return 0 }
        return (deepSleepMinutes / totalSleepMinutes) * 100.0
    }

    var remSleepPercentage: Double {
        guard totalSleepMinutes > 0 else { return 0 }
        return (remSleepMinutes / totalSleepMinutes) * 100.0
    }

    var lightSleepPercentage: Double {
        guard totalSleepMinutes > 0 else { return 0 }
        return (lightSleepMinutes / totalSleepMinutes) * 100.0
    }

    var formattedDuration: String {
        let hours = Int(totalSleepMinutes) / 60
        let mins = Int(totalSleepMinutes) % 60
        return "\(hours)h \(mins)m"
    }
}

File: android/core/model/src/main/kotlin/com/apexfit/core/model/SleepSession.kt
package com.apexfit.core.model

data class SleepSession(
    val id: String = java.util.UUID.randomUUID().toString(),
    val startDate: Long,
    val endDate: Long,
    val isMainSleep: Boolean = true,
    val isNap: Boolean = false,
    val totalSleepMinutes: Double = 0.0,
    val timeInBedMinutes: Double = (endDate - startDate) / 60_000.0,
    val lightSleepMinutes: Double = 0.0,
    val deepSleepMinutes: Double = 0.0,
    val remSleepMinutes: Double = 0.0,
    val awakeMinutes: Double = 0.0,
    val awakenings: Int = 0,
    val sleepOnsetLatencyMinutes: Double? = null,
    val sleepEfficiency: Double = 0.0,
    val sleepPerformance: Double? = null,
    val sleepNeedHours: Double? = null,
    val createdAt: Long = System.currentTimeMillis(),
    val dailyMetricId: String? = null,
    val stages: List<SleepStage> = emptyList(),
) {
    val totalSleepHours: Double get() = totalSleepMinutes / 60.0

    val deepSleepPercentage: Double
        get() = if (totalSleepMinutes > 0) (deepSleepMinutes / totalSleepMinutes) * 100.0 else 0.0

    val remSleepPercentage: Double
        get() = if (totalSleepMinutes > 0) (remSleepMinutes / totalSleepMinutes) * 100.0 else 0.0

    val lightSleepPercentage: Double
        get() = if (totalSleepMinutes > 0) (lightSleepMinutes / totalSleepMinutes) * 100.0 else 0.0

    val formattedDuration: String
        get() {
            val hours = totalSleepMinutes.toInt() / 60
            val mins = totalSleepMinutes.toInt() % 60
            return "${hours}h ${mins}m"
        }
}
